File: backend/__SLUG___node/src/domains/auth/datasource.ts
import EmailPassword from "supertokens-node/recipe/emailpassword";
import Session from "supertokens-node/recipe/session";
import UserRoles from "supertokens-node/recipe/userroles";
import supertokens from "supertokens-node";
import { v4 as uuidv4 } from "uuid";
import type { Database } from "../../logger/db";
import * as pagination from "../../shared/pagination/pagination";
import {
  type ApplicationUser,
  EmailAlreadyExistsError,
  type User,
  UserNotFoundError,
  WrongCredentialsError,
} from "./models";

/** Maps API field names to DB column names for users. */
export const userColumnMap: Record<string, string> = {
  displayName: "display_name",
  email: "email",
  isDisabled: "is_disabled",
  dateCreated: "date_created",
};

interface UserRow {
  id: string;
  supertokens_id: string;
  email: string;
  display_name: string;
  is_disabled: boolean;
  date_created: string;
  last_modified: string;
}

function rowToUser(row: UserRow): ApplicationUser {
  return {
    id: row.id,
    supertokensId: row.supertokens_id,
    email: row.email,
    displayName: row.display_name,
    isDisabled: row.is_disabled,
    roles: [],
    dateCreated: row.date_created,
    lastModified: row.last_modified,
  };
}

/** Handles auth interactions: SuperTokens for credentials, postgres for application users. */
export class Datasource {
  constructor(private readonly db: Database) {}

  /** Creates a new user via SuperTokens. */
  async signUp(email: string, password: string): Promise<string> {
    const resp = await EmailPassword.signUp("public", email, password);
    if (resp.status === "EMAIL_ALREADY_EXISTS_ERROR") {
      throw new EmailAlreadyExistsError();
    }
    return resp.user.id;
  }

  /** Authenticates a user via SuperTokens. */
  async signIn(email: string, password: string): Promise<string> {
    const resp = await EmailPassword.signIn("public", email, password);
    if (resp.status === "WRONG_CREDENTIALS_ERROR") {
      throw new WrongCredentialsError();
    }
    return resp.user.id;
  }

  /** Retrieves SuperTokens user information. */
  async getUser(userId: string): Promise<User> {
    const userInfo = await supertokens.getUser(userId);
    if (!userInfo) {
      throw new UserNotFoundError();
    }
    return {
      id: userInfo.id,
      email: userInfo.emails[0] ?? "",
      name: userInfo.emails[0] ?? "",
      roles: [{ name: "admin" }],
    };
  }

  /** Inserts a new row into the users table and assigns roles. */
  async createApplicationUser(
    supertokensId: string,
    email: string,
    displayName: string,
    roles: string[],
  ): Promise<ApplicationUser> {
    const now = new Date().toISOString();
    const id = uuidv4();
    await this.db.query(
      `INSERT INTO users (id, supertokens_id, email, display_name, date_created, last_modified)
       VALUES ($1, $2, $3, $4, $5, $6)`,
      [id, supertokensId, email, displayName, now, now],
    );
    for (const role of roles) {
      await UserRoles.addRoleToUser("public", supertokensId, role);
    }
    return {
      id,
      supertokensId,
      email,
      displayName,
      isDisabled: false,
      roles,
      dateCreated: now,
      lastModified: now,
    };
  }

  /** Looks up the application user by SuperTokens user ID. */
  async getApplicationUserBySupertokensId(supertokensId: string): Promise<ApplicationUser> {
    const result = await this.db.query<UserRow>(
      `SELECT id, supertokens_id, email, display_name, is_disabled, date_created, last_modified
       FROM users WHERE supertokens_id = $1`,
      [supertokensId],
    );
    if (result.rowCount === 0) {
      throw new UserNotFoundError();
    }
    return rowToUser(result.rows[0]);
  }

  /** Looks up an application user by their application ID. */
  async getApplicationUserById(id: string): Promise<ApplicationUser> {
    const result = await this.db.query<UserRow>(
      `SELECT id, supertokens_id, email, display_name, is_disabled, date_created, last_modified
       FROM users WHERE id = $1`,
      [id],
    );
    if (result.rowCount === 0) {
      throw new UserNotFoundError();
    }
    return rowToUser(result.rows[0]);
  }

  /** Verifies the old password then updates it via SuperTokens. */
  async changePassword(email: string, oldPassword: string, newPassword: string): Promise<void> {
    const signInResp = await EmailPassword.signIn("public", email, oldPassword);
    if (signInResp.status === "WRONG_CREDENTIALS_ERROR") {
      throw new WrongCredentialsError();
    }
    const userId = signInResp.user.id;
    const updateResp = await EmailPassword.updateEmailOrPassword({
      recipeUserId: signInResp.user.loginMethods[0].recipeUserId,
      password: newPassword,
    });
    if (updateResp.status === "UNKNOWN_USER_ID_ERROR") {
      throw new UserNotFoundError();
    }
    void userId;
  }

  /** Returns a paged list of application users matching the given params. */
  async listUsers(p: pagination.ResolvedParams) {
    const args: unknown[] = [];
    let argIdx = 1;
    const conditions: string[] = [];

    for (const f of p.filters) {
      switch (f.op) {
        case "eq":
          conditions.push(`${f.column} = $${argIdx++}`);
          args.push(f.value);
          break;
        case "contains":
          conditions.push(`${f.column} ILIKE $${argIdx++}`);
          args.push(`%${f.value}%`);
          break;
        case "gt":
          conditions.push(`${f.column} > $${argIdx++}`);
          args.push(f.value);
          break;
        case "gte":
          conditions.push(`${f.column} >= $${argIdx++}`);
          args.push(f.value);
          break;
        case "lt":
          conditions.push(`${f.column} < $${argIdx++}`);
          args.push(f.value);
          break;
        case "lte":
          conditions.push(`${f.column} <= $${argIdx++}`);
          args.push(f.value);
          break;
      }
    }

    const where = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
    const sortCol = p.sort || "display_name";
    const orderBy = `ORDER BY ${sortCol} ${p.sortDir.toUpperCase()}`;

    const countResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) FROM users ${where}`,
      args,
    );
    const total = Number(countResult.rows[0].count);

    const dataArgs = [...args, p.pageSize, pagination.offset(p)];
    const dataQuery = `SELECT id, supertokens_id, email, display_name, is_disabled, date_created, last_modified
       FROM users ${where} ${orderBy} LIMIT $${argIdx} OFFSET $${argIdx + 1}`;
    const dataResult = await this.db.query<UserRow>(dataQuery, dataArgs);

    return {
      data: dataResult.rows.map(rowToUser),
      total,
      page: p.page,
      pageSize: p.pageSize,
    };
  }

  /** Returns the role names assigned to a SuperTokens user. */
  async getRolesForUser(supertokensId: string): Promise<string[]> {
    const resp = await UserRoles.getRolesForUser("public", supertokensId);
    return resp.roles;
  }

  /** Updates email, display_name, and roles for a user. */
  async updateUser(
    id: string,
    email: string,
    displayName: string,
    roles: string[],
  ): Promise<ApplicationUser> {
    const now = new Date().toISOString();
    const result = await this.db.query<UserRow>(
      `UPDATE users SET email = $1, display_name = $2, last_modified = $3
       WHERE id = $4
       RETURNING id, supertokens_id, email, display_name, is_disabled, date_created, last_modified`,
      [email, displayName, now, id],
    );
    if (result.rowCount === 0) {
      throw new UserNotFoundError();
    }
    const u = rowToUser(result.rows[0]);

    const currentRolesResp = await UserRoles.getRolesForUser("public", u.supertokensId);
    const currentRoles = new Set(currentRolesResp.roles);
    const desiredRoles = new Set(roles);

    for (const r of desiredRoles) {
      if (!currentRoles.has(r)) {
        await UserRoles.addRoleToUser("public", u.supertokensId, r);
      }
    }
    for (const r of currentRoles) {
      if (!desiredRoles.has(r)) {
        await UserRoles.removeUserRole("public", u.supertokensId, r);
      }
    }
    u.roles = roles;
    return u;
  }

  /** Resets a user's password without requiring the old password. */
  async adminResetPassword(supertokensId: string, newPassword: string): Promise<void> {
    const userInfo = await supertokens.getUser(supertokensId);
    if (!userInfo) {
      throw new UserNotFoundError();
    }
    const resp = await EmailPassword.updateEmailOrPassword({
      recipeUserId: userInfo.loginMethods[0].recipeUserId,
      password: newPassword,
    });
    if (resp.status === "UNKNOWN_USER_ID_ERROR") {
      throw new UserNotFoundError();
    }
  }

  /** Sets is_disabled=true and revokes all active sessions. */
  async disableUser(id: string): Promise<void> {
    const now = new Date().toISOString();
    const result = await this.db.query<{ supertokens_id: string }>(
      `UPDATE users SET is_disabled = true, last_modified = $1 WHERE id = $2 RETURNING supertokens_id`,
      [now, id],
    );
    if (result.rowCount === 0) {
      throw new UserNotFoundError();
    }
    await Session.revokeAllSessionsForUser(result.rows[0].supertokens_id);
  }

  /** Sets is_disabled=false. */
  async enableUser(id: string): Promise<void> {
    const now = new Date().toISOString();
    const result = await this.db.query(
      `UPDATE users SET is_disabled = false, last_modified = $1 WHERE id = $2`,
      [now, id],
    );
    if (result.rowCount === 0) {
      throw new UserNotFoundError();
    }
  }

  /** Removes a SuperTokens user (best-effort, used for signup rollback). */
  async deleteSupertokensUser(supertokensId: string): Promise<void> {
    try {
      await supertokens.deleteUser(supertokensId);
    } catch {
      // best-effort cleanup
    }
  }
}

File: backend/__SLUG__/src/domains/auth/datasource.go
package auth

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/supertokens/supertokens-golang/recipe/emailpassword"
	epmodels "github.com/supertokens/supertokens-golang/recipe/emailpassword/epmodels"
	"github.com/supertokens/supertokens-golang/recipe/session"
	"github.com/supertokens/supertokens-golang/recipe/userroles"
	userrolesmodels "github.com/supertokens/supertokens-golang/recipe/userroles/userrolesmodels"
	"github.com/supertokens/supertokens-golang/supertokens"
	"go_backend/src/logger"
	"go_backend/src/shared/pagination"
)

// userColumnMap maps API field names to DB column names for users.
var userColumnMap = map[string]string{
	"displayName": "display_name",
	"email":       "email",
	"isDisabled":  "is_disabled",
	"dateCreated": "date_created",
}

type emailPasswordAPI interface {
	SignUp(tenantID, email, password string) (epmodels.SignUpResponse, error)
	SignIn(tenantID, email, password string) (epmodels.SignInResponse, error)
	GetUserByID(userID string) (*epmodels.User, error)
	UpdateEmailOrPassword(userID string, email, password *string, applyPasswordPolicy *bool, tenantIDForPasswordPolicy *string) (epmodels.UpdateEmailOrPasswordResponse, error)
}

type sessionAPI interface {
	RevokeAllSessionsForUser(userID string, tenantID *string) ([]string, error)
}

type userRolesAPI interface {
	AddRoleToUser(tenantID, userID, role string) (userrolesmodels.AddRoleToUserResponse, error)
	GetRolesForUser(tenantID, userID string) (userrolesmodels.GetRolesForUserResponse, error)
	RemoveUserRole(tenantID, userID, role string) (userrolesmodels.RemoveUserRoleResponse, error)
}

type supertokensAPI interface {
	DeleteUser(userID string) error
}

type realEP struct{}

func (r *realEP) SignUp(tenantID, email, password string) (epmodels.SignUpResponse, error) {
	return emailpassword.SignUp(tenantID, email, password)
}
func (r *realEP) SignIn(tenantID, email, password string) (epmodels.SignInResponse, error) {
	return emailpassword.SignIn(tenantID, email, password)
}
func (r *realEP) GetUserByID(userID string) (*epmodels.User, error) {
	return emailpassword.GetUserByID(userID)
}
func (r *realEP) UpdateEmailOrPassword(userID string, email, password *string, apply *bool, tenantID *string) (epmodels.UpdateEmailOrPasswordResponse, error) {
	return emailpassword.UpdateEmailOrPassword(userID, email, password, apply, tenantID)
}

type realSession struct{}

func (r *realSession) RevokeAllSessionsForUser(userID string, tenantID *string) ([]string, error) {
	return session.RevokeAllSessionsForUser(userID, tenantID)
}

type realUserRoles struct{}

func (r *realUserRoles) AddRoleToUser(tenantID, userID, role string) (userrolesmodels.AddRoleToUserResponse, error) {
	return userroles.AddRoleToUser(tenantID, userID, role)
}
func (r *realUserRoles) GetRolesForUser(tenantID, userID string) (userrolesmodels.GetRolesForUserResponse, error) {
	return userroles.GetRolesForUser(tenantID, userID)
}
func (r *realUserRoles) RemoveUserRole(tenantID, userID, role string) (userrolesmodels.RemoveUserRoleResponse, error) {
	return userroles.RemoveUserRole(tenantID, userID, role)
}

type realST struct{}

func (r *realST) DeleteUser(userID string) error { return supertokens.DeleteUser(userID) }

// Datasource handles auth interactions: SuperTokens for credentials, postgres for application users.
type Datasource struct {
	db    logger.Database
	ep    emailPasswordAPI
	sess  sessionAPI
	roles userRolesAPI
	st    supertokensAPI
}

// NewDatasource creates a new datasource instance.
func NewDatasource(db logger.Database) *Datasource {
	return &Datasource{
		db:    db,
		ep:    &realEP{},
		sess:  &realSession{},
		roles: &realUserRoles{},
		st:    &realST{},
	}
}

// SignUp creates a new user via SuperTokens.
func (ds *Datasource) SignUp(email, password string) (string, error) {
	resp, err := ds.ep.SignUp("public", email, password)
	if err != nil {
		return "", err
	}
	if resp.EmailAlreadyExistsError != nil {
		return "", &EmailAlreadyExistsError{}
	}
	return resp.OK.User.ID, nil
}

// SignIn authenticates a user via SuperTokens.
func (ds *Datasource) SignIn(email, password string) (string, error) {
	resp, err := ds.ep.SignIn("public", email, password)
	if err != nil {
		return "", err
	}
	if resp.WrongCredentialsError != nil {
		return "", &WrongCredentialsError{}
	}
	return resp.OK.User.ID, nil
}

// GetUser retrieves SuperTokens user information.
func (ds *Datasource) GetUser(userID string) (*User, error) {
	userInfo, err := ds.ep.GetUserByID(userID)
	if err != nil {
		return nil, err
	}
	if userInfo == nil {
		return nil, &UserNotFoundError{}
	}
	return &User{
		ID:    userInfo.ID,
		Email: userInfo.Email,
		Name:  userInfo.Email,
		Roles: []UserRole{{Name: "admin"}},
	}, nil
}

// CreateApplicationUser inserts a new row into the users table and assigns roles.
func (ds *Datasource) CreateApplicationUser(ctx context.Context, supertokensID, email, displayName string, roles []string) (*ApplicationUser, error) {
	now := time.Now().UTC().Format(time.RFC3339)
	id := uuid.New().String()
	_, err := ds.db.ExecContext(ctx,
		`INSERT INTO users (id, supertokens_id, email, display_name, date_created, last_modified)
		 VALUES ($1, $2, $3, $4, $5, $6)`,
		id, supertokensID, email, displayName, now, now,
	)
	if err != nil {
		return nil, fmt.Errorf("insert application user: %w", err)
	}
	for _, role := range roles {
		if _, err := ds.roles.AddRoleToUser("public", supertokensID, role); err != nil {
			return nil, fmt.Errorf("add role %s: %w", role, err)
		}
	}
	return &ApplicationUser{
		ID:            id,
		SupertokensID: supertokensID,
		Email:         email,
		DisplayName:   displayName,
		IsDisabled:    false,
		Roles:         roles,
		DateCreated:   now,
		LastModified:  now,
	}, nil
}

// GetApplicationUserBySupertokensID looks up the application user by SuperTokens user ID.
func (ds *Datasource) GetApplicationUserBySupertokensID(ctx context.Context, supertokensID string) (*ApplicationUser, error) {
	row := ds.db.QueryRowContext(ctx,
		`SELECT id, supertokens_id, email, display_name, is_disabled, date_created, last_modified
		 FROM users WHERE supertokens_id = $1`, supertokensID)
	u := &ApplicationUser{}
	if err := row.Scan(&u.ID, &u.SupertokensID, &u.Email, &u.DisplayName, &u.IsDisabled, &u.DateCreated, &u.LastModified); err != nil {
		if err == sql.ErrNoRows {
			return nil, &UserNotFoundError{}
		}
		return nil, fmt.Errorf("get application user: %w", err)
	}
	return u, nil
}

// ChangePassword verifies the old password then updates it via SuperTokens.
func (ds *Datasource) ChangePassword(ctx context.Context, email, oldPassword, newPassword string) error {
	// Verify old password
	signInResp, err := ds.ep.SignIn("public", email, oldPassword)
	if err != nil {
		return err
	}
	if signInResp.WrongCredentialsError != nil {
		return &WrongCredentialsError{}
	}
	userID := signInResp.OK.User.ID
	// Update password
	updateResp, err := ds.ep.UpdateEmailOrPassword(userID, nil, &newPassword, nil, nil)
	if err != nil {
		return fmt.Errorf("update password: %w", err)
	}
	if updateResp.UnknownUserIdError != nil {
		return &UserNotFoundError{}
	}
	return nil
}

// ListUsers returns a paged list of application users matching the given params.
func (ds *Datasource) ListUsers(ctx context.Context, p pagination.ResolvedParams) (pagination.PagedResponse[ApplicationUser], error) {
	var args []interface{}
	argIdx := 1

	// WHERE clause
	var conditions []string
	for _, f := range p.Filters {
		switch f.Op {
		case pagination.OpEq:
			conditions = append(conditions, fmt.Sprintf("%s = $%d", f.Column, argIdx))
			args = append(args, f.Value)
			argIdx++
		case pagination.OpContains:
			conditions = append(conditions, fmt.Sprintf("%s ILIKE $%d", f.Column, argIdx))
			args = append(args, "%"+f.Value+"%")
			argIdx++
		case pagination.OpGt:
			conditions = append(conditions, fmt.Sprintf("%s > $%d", f.Column, argIdx))
			args = append(args, f.Value)
			argIdx++
		case pagination.OpGte:
			conditions = append(conditions, fmt.Sprintf("%s >= $%d", f.Column, argIdx))
			args = append(args, f.Value)
			argIdx++
		case pagination.OpLt:
			conditions = append(conditions, fmt.Sprintf("%s < $%d", f.Column, argIdx))
			args = append(args, f.Value)
			argIdx++
		case pagination.OpLte:
			conditions = append(conditions, fmt.Sprintf("%s <= $%d", f.Column, argIdx))
			args = append(args, f.Value)
			argIdx++
		}
	}

	where := ""
	if len(conditions) > 0 {
		where = "WHERE " + strings.Join(conditions, " AND ")
	}

	// ORDER BY
	sortCol := "display_name"
	if p.Sort != "" {
		sortCol = p.Sort
	}
	orderBy := fmt.Sprintf("ORDER BY %s %s", sortCol, strings.ToUpper(string(p.SortDir)))

	// COUNT
	countQuery := fmt.Sprintf(`SELECT COUNT(*) FROM users %s`, where)
	var total int
	if err := ds.db.QueryRowContext(ctx, countQuery, args...).Scan(&total); err != nil {
		return pagination.PagedResponse[ApplicationUser]{}, fmt.Errorf("count users: %w", err)
	}

	// DATA
	dataQuery := fmt.Sprintf(
		`SELECT id, supertokens_id, email, display_name, is_disabled, date_created, last_modified
		 FROM users %s %s LIMIT $%d OFFSET $%d`,
		where, orderBy, argIdx, argIdx+1,
	)
	args = append(args, p.PageSize, p.Offset())

	rows, err := ds.db.QueryContext(ctx, dataQuery, args...)
	if err != nil {
		return pagination.PagedResponse[ApplicationUser]{}, fmt.Errorf("list users: %w", err)
	}
	defer rows.Close()

	var users []ApplicationUser
	for rows.Next() {
		var u ApplicationUser
		if err := rows.Scan(&u.ID, &u.SupertokensID, &u.Email, &u.DisplayName, &u.IsDisabled, &u.DateCreated, &u.LastModified); err != nil {
			return pagination.PagedResponse[ApplicationUser]{}, fmt.Errorf("scan user: %w", err)
		}
		users = append(users, u)
	}
	if err := rows.Err(); err != nil {
		return pagination.PagedResponse[ApplicationUser]{}, err
	}
	if users == nil {
		users = []ApplicationUser{}
	}
	return pagination.PagedResponse[ApplicationUser]{
		Data:     users,
		Total:    total,
		Page:     p.Page,
		PageSize: p.PageSize,
	}, nil
}

// GetRolesForUser returns the role names assigned to a SuperTokens user.
func (ds *Datasource) GetRolesForUser(ctx context.Context, supertokensID string) ([]string, error) {
	resp, err := ds.roles.GetRolesForUser("public", supertokensID)
	if err != nil {
		return nil, fmt.Errorf("get roles for user: %w", err)
	}
	return resp.OK.Roles, nil
}

// GetApplicationUserByID looks up an application user by their application ID.
func (ds *Datasource) GetApplicationUserByID(ctx context.Context, id string) (*ApplicationUser, error) {
	row := ds.db.QueryRowContext(ctx,
		`SELECT id, supertokens_id, email, display_name, is_disabled, date_created, last_modified
		 FROM users WHERE id = $1`, id)
	u := &ApplicationUser{}
	if err := row.Scan(&u.ID, &u.SupertokensID, &u.Email, &u.DisplayName, &u.IsDisabled, &u.DateCreated, &u.LastModified); err != nil {
		if err == sql.ErrNoRows {
			return nil, &UserNotFoundError{}
		}
		return nil, fmt.Errorf("get application user by id: %w", err)
	}
	return u, nil
}

// UpdateUser updates email, display_name, and roles for a user.
func (ds *Datasource) UpdateUser(ctx context.Context, id, email, displayName string, roles []string) (*ApplicationUser, error) {
	now := time.Now().UTC().Format(time.RFC3339)
	row := ds.db.QueryRowContext(ctx,
		`UPDATE users SET email = $1, display_name = $2, last_modified = $3
		 WHERE id = $4
		 RETURNING id, supertokens_id, email, display_name, is_disabled, date_created, last_modified`,
		email, displayName, now, id)
	u := &ApplicationUser{}
	if err := row.Scan(&u.ID, &u.SupertokensID, &u.Email, &u.DisplayName, &u.IsDisabled, &u.DateCreated, &u.LastModified); err != nil {
		if err == sql.ErrNoRows {
			return nil, &UserNotFoundError{}
		}
		return nil, fmt.Errorf("update user: %w", err)
	}

	// Diff roles
	currentResp, err := ds.roles.GetRolesForUser("public", u.SupertokensID)
	if err != nil {
		return nil, fmt.Errorf("get current roles: %w", err)
	}
	currentRoles := make(map[string]bool)
	for _, r := range currentResp.OK.Roles {
		currentRoles[r] = true
	}
	desiredRoles := make(map[string]bool)
	for _, r := range roles {
		desiredRoles[r] = true
	}
	for r := range desiredRoles {
		if !currentRoles[r] {
			if _, err := ds.roles.AddRoleToUser("public", u.SupertokensID, r); err != nil {
				return nil, fmt.Errorf("add role %s: %w", r, err)
			}
		}
	}
	for r := range currentRoles {
		if !desiredRoles[r] {
			if _, err := ds.roles.RemoveUserRole("public", u.SupertokensID, r); err != nil {
				return nil, fmt.Errorf("remove role %s: %w", r, err)
			}
		}
	}
	u.Roles = roles
	return u, nil
}

// AdminResetPassword resets a user's password without requiring the old password.
func (ds *Datasource) AdminResetPassword(ctx context.Context, supertokensID, newPassword string) error {
	resp, err := ds.ep.UpdateEmailOrPassword(supertokensID, nil, &newPassword, nil, nil)
	if err != nil {
		return fmt.Errorf("reset password: %w", err)
	}
	if resp.UnknownUserIdError != nil {
		return &UserNotFoundError{}
	}
	return nil
}

// DisableUser sets is_disabled=true and revokes all active sessions.
func (ds *Datasource) DisableUser(ctx context.Context, id string) error {
	now := time.Now().UTC().Format(time.RFC3339)
	row := ds.db.QueryRowContext(ctx,
		`UPDATE users SET is_disabled = true, last_modified = $1 WHERE id = $2 RETURNING supertokens_id`,
		now, id)
	var supertokensID string
	if err := row.Scan(&supertokensID); err != nil {
		if err == sql.ErrNoRows {
			return &UserNotFoundError{}
		}
		return fmt.Errorf("disable user: %w", err)
	}
	if _, err := ds.sess.RevokeAllSessionsForUser(supertokensID, nil); err != nil {
		return fmt.Errorf("revoke sessions: %w", err)
	}
	return nil
}

// EnableUser sets is_disabled=false.
func (ds *Datasource) EnableUser(ctx context.Context, id string) error {
	now := time.Now().UTC().Format(time.RFC3339)
	result, err := ds.db.ExecContext(ctx,
		`UPDATE users SET is_disabled = false, last_modified = $1 WHERE id = $2`, now, id)
	if err != nil {
		return fmt.Errorf("enable user: %w", err)
	}
	rows, _ := result.RowsAffected()
	if rows == 0 {
		return &UserNotFoundError{}
	}
	return nil
}

// DeleteSupertokensUser removes a SuperTokens user (used for signup rollback).
func (ds *Datasource) DeleteSupertokensUser(supertokensID string) {
	// Best-effort cleanup — ignore error
	_ = ds.st.DeleteUser(supertokensID)
}

// Custom errors

type EmailAlreadyExistsError struct{}

func (e *EmailAlreadyExistsError) Error() string { return "email already exists" }

type WrongCredentialsError struct{}

func (e *WrongCredentialsError) Error() string { return "wrong credentials" }

type UserNotFoundError struct{}

func (e *UserNotFoundError) Error() string { return "user not found" }

type SelfDisableError struct{}

func (e *SelfDisableError) Error() string { return "cannot disable your own account" }
